### DISCUSSION TITLE
Error while using browser navigation: Unhandled Runtime Error TypeError: Cannot read property 'indexOf' of undefined

### DISCUSSION THREAD

**Discussion by antonioOrtiz** [vercel/next.js#13010] [Help] [answered]
Ahoy!

I am on  `"next": "^9.4.0",` and simply when I start in dev mode and use the browser forward button  I get this error:

**Unhandled Runtime Error**
```
TypeError: Cannot read property 'indexOf' of undefined

Call Stack

addBasePath
  webpack-internal:///./node_modules/next/dist/next-server/lib/router/router.js (38:15)
eval
  webpack-internal:///./node_modules/next/dist/next-server/lib/router/router.js (331:15)
new Promise
  <anonymous>
Router.change
  webpack-internal:///./node_modules/next/dist/next-server/lib/router/router.js (317:14)
Router.replace
  webpack-internal:///./node_modules/next/dist/next-server/lib/router/router.js (310:19)
Router.onPopState
  webpack-internal:///./node_modules/next/dist/next-server/lib/router/router.js (169:13)
```
Little backstory, I am using `react-router` because the `next-router` was not working with a `semantic-ui-react` use case, but considering the errors are pointing to the 'next' package in the `node_modules` I figure that's not the issue.

Appreciate any insight!

**Comment by febinsathar** [comment]
+1

**Comment by timneutkens** [comment]
Can you provide a full reproduction? otherwise it's hard for people to look into.

**Reply by antonioOrtiz** [reply]
Do we have a guide of what I should include in the reproduction?

**Reply by timneutkens** [reply]
A complete application to run

**Reply by antonioOrtiz** [reply]
Great. Thank you. Here is a link to my [repo](https://github.com/antonioOrtiz/hillfinder?files=1) or should it be hosted?

**Reply by malixsys** [reply]
Got the same @antonioOrtiz


@timneutkens When you use `react-router` and hit the back button, `path` is undefined
`window.addEventListener('popstate', this.onPopState)` fires at https://github.com/vercel/next.js/blob/canary/packages/next/next-server/lib/router/router.ts#L270
Sadly, 
```
export function addBasePath(path: string): string {
  return `${basePath}${path}`
}
```
at https://github.com/vercel/next.js/blob/canary/packages/next/next-server/lib/router/router.ts#L23
is "compiled" as 
```
function addBasePath(path){return path.indexOf(basePath)!==0?basePath+path:path;}
```
So it blows up... I suggest adding a default:
```
export function addBasePath(path: string = ''): string {
  return `${basePath}${path}`
}
```

**Reply by antonioOrtiz** [reply]
@timneutkens Thanks for responding! However I think my version of next might be different from the above link, when I did a global search for `addBasePath` using vscode I found:

![Screen Shot 2020-06-27 at 1 59 38 PM](https://user-images.githubusercontent.com/4282013/85928857-51716f00-b87e-11ea-803b-258e26725e72.jpg)

That being as it is, I found this:

```
/// <reference types="node" />
import { ParsedUrlQuery } from 'querystring';
import { ComponentType } from 'react';
import { UrlObject } from 'url';
import { MittEmitter } from '../mitt';
import { NextPageContext } from '../utils';
export declare function addBasePath(path: string): string;  <--- Found it here.
export declare function delBasePath(path: string): string:
```

As you can see it's different, So I added this:

```
export declare function addBasePath(path: string = ''): string {
  return `${basePath}${path}`
}
```

Now I'm getting:

```
Unhandled Runtime Error
TypeError: Cannot read property 'originalUrl' of undefined

Call Stack
_callee$
webpack-internal:///./with-react-router/with-react-router.js (97:95)
tryCatch
webpack-internal:///./node_modules/regenerator-runtime/runtime.js (62:40)
Generator.invoke [as _invoke]
webpack-internal:///./node_modules/regenerator-runtime/runtime.js (288:22)
Generator.prototype.<computed> [as next]
webpack-internal:///./node_modules/regenerator-runtime/runtime.js (114:21)
asyncGeneratorStep
webpack-internal:///./node_modules/@babel/runtime/helpers/esm/asyncToGenerator.js (5:24)
_next
webpack-internal:///./node_modules/@babel/runtime/helpers/esm/asyncToGenerator.js (27:9)
eval
webpack-internal:///./node_modules/@babel/runtime/helpers/esm/asyncToGenerator.js (34:7)
new Promise
<anonymous>
Function.eval
webpack-internal:///./node_modules/@babel/runtime/helpers/esm/asyncToGenerator.js (23:12)
Function.getInitialProps
webpack-internal:///./with-react-router/with-react-router.js (112:35)
eval
webpack-internal:///./node_modules/next-redux-wrapper/es6/index.js (140:66)
step
webpack-internal:///./node_modules/next-redux-wrapper/es6/index.js (58:23)
Object.eval [as next]
webpack-internal:///./node_modules/next-redux-wrapper/es6/index.js (39:53)
eval
webpack-internal:///./node_modules/next-redux-wrapper/es6/index.js (33:71)
new Promise
<anonymous>
__awaiter
webpack-internal:///./node_modules/next-redux-wrapper/es6/index.js (29:12)
Function._a.getInitialProps
webpack-internal:///./node_modules/next-redux-wrapper/es6/index.js (120:57)
_callee$
webpack-internal:///./node_modules/next/dist/next-server/lib/utils.js (108:24)
tryCatch
webpack-internal:///./node_modules/next/node_modules/regenerator-runtime/runtime.js (45:40)
Generator.invoke [as _invoke]
webpack-internal:///./node_modules/next/node_modules/regenerator-runtime/runtime.js (274:22)
Generator.prototype.<computed> [as next]
webpack-internal:///./node_modules/next/node_modules/regenerator-runtime/runtime.js (97:21)
asyncGeneratorStep
webpack-internal:///./node_modules/next/node_modules/@babel/runtime/helpers/asyncToGenerator.js (3:24)
_next
webpack-internal:///./node_modules/next/node_modules/@babel/runtime/helpers/asyncToGenerator.js (25:9)
eval
webpack-internal:///./node_modules/next/node_modules/@babel/runtime/helpers/asyncToGenerator.js (32:7)
new Promise
<anonymous>
eval
webpack-internal:///./node_modules/next/node_modules/@babel/runtime/helpers/asyncToGenerator.js (21:12)
_loadGetInitialProps
webpack-internal:///./node_modules/next/dist/next-server/lib/utils.js (145:31)
loadGetInitialProps
webpack-internal:///./node_modules/next/dist/next-server/lib/utils.js (59:31)
Router.getInitialProps
webpack-internal:///./node_modules/next/dist/next-server/lib/router/router.js (797:45)
eval
webpack-internal:///./node_modules/next/dist/next-server/lib/router/router.js (575:101)
Router._getData
webpack-internal:///./node_modules/next/dist/next-server/lib/router/router.js (775:14)
eval
webpack-internal:///./node_modules/next/dist/next-server/lib/router/router.js (574:23)
```

Hey this is my [repo](https://github.com/antonioOrtiz/hillfinder):

And these are the relevant files

[with-react-router.js](https://github.com/antonioOrtiz/hillfinder/tree/hillfinder-userAvatar/with-react-router)
[server/index.js](https://github.com/antonioOrtiz/hillfinder/blob/hillfinder-userAvatar/server/index.js)
[_app.js](https://github.com/antonioOrtiz/hillfinder/blob/hillfinder-userAvatar/pages/_app.jsx)

I'd appreciate you or anyone for that matter shedding some light!

**Comment by jtorresdev** [comment]
@antonioOrtiz  verify that `<Link />` has the prop "href"

**Reply by antonioOrtiz** [reply]
Just realized I forgot to mention I am using React-router instead of the built-in Next router,  because I had a problem implementing Next Router with a navigation supplied by Semantic-UI-React.

And being the errors are all being produced from next-server I wonder if there is a conflict with react-router? Is there a way to explicitly shut off next-router? Like an escape hatch?

**Reply by malixsys** [reply]
@antonioOrtiz I fixed it by using:
```
export default withRouter(App);
```

**Reply by malixsys** [reply]
Here's my fixed `index.js`:
```javascript
// /pages/index.js
import React, { lazy, Suspense } from 'react';
import { withRouter } from 'next/router';
import { BrowserRouter as Router } from 'react-router-dom';
import Loader from '../src/client/components/Loader';

const AppContent = lazy(() => import('../src/client/components/App/AppContent'));

const App = () => {
  return (
      <Router>
        <Suspense fallback={<Loader />}>
          <AppContent />
        </Suspense>
      </Router>
  );
};

export default withRouter(App);
```

**Reply by malixsys** [reply]
CC @NachoJusticia @notiles

**Reply by malixsys** [reply]
CC @cedie712

**Comment by NachoJusticia** [comment]
Hi, I am having the same problem.

I am using `next` v9.4.0, `react` v16.13.1, `reac-router-dom` v5.1.2 and `history` v4.10.1`.

The `history` dependency is used to create the browser history using `createBrowserHistory` function.
I am using nextjs file system based routing for navigating between sign in, sign up etc and "dashboard" pages, and also `react-router-dom` with `BrowseRouter` for navigation inside the "dashboard" page.

I have tried with both `replace` and `push` functions of `next/router` to navigate from `404` nextjs static page to the "dashboard", and it works fine. Anyhow, when I keep navigating between "dashboard" pages sometimes I got the error `Cannot read property 'indexOf' of undefined` when navigating back.

```
Unhandled Runtime Error
TypeError: Cannot read property 'indexOf' of undefined

Call Stack
addBasePath
../../../../next-server/lib/router/router.ts (22:9)
eval
../../../../next-server/lib/router/router.ts (382:12)
new Promise
<anonymous>
Router.change
../../../../next-server/lib/router/router.ts (368:11)
Router.replace
../../../../next-server/lib/router/router.ts (359:11)
Router.onPopState
../../../../next-server/lib/router/router.ts (303:4)
```

**Comment by cedrickvstheworld** [comment]
To somebody who will stumble with this issue and you just want some quick work-around: I have faced the same issue with the same version. I step back to version 9.1.7 and it's gone. After that, I faced another problem with css imports. For that, you have to install builder stuff (that's what they call it i guess?). Here's a link that will guide you https://github.com/zeit/next-plugins/tree/master/packages/next-css

**Comment by ShubhamKushwah** [comment]
Apparently this issue shows up for me, if I use Link from `next/link`, and didn't specify `href`.

So, just try doing
```
<Link href="#"><a>Home</a></Link>
```
to see if it works.

**Comment by notiles** [comment]
Have got the same error mentioned by @antonioOrtiz .

"next": "^9.4.2",
"react": "^16.13.1",

Click on a Link (with href & as) click previous on browser navigation.

```
Unhandled Runtime Error
TypeError: Cannot read property 'indexOf' of undefined

Call Stack
addBasePath
../../../../next-server/lib/router/router.ts (22:9)
eval
../../../../next-server/lib/router/router.ts (382:12)
new Promise
<anonymous>
Router.change
../../../../next-server/lib/router/router.ts (368:11)
Router.replace
../../../../next-server/lib/router/router.ts (359:11)
Router.onPopState
../../../../next-server/lib/router/router.ts (303:4)
```

**Comment by antonioOrtiz** [comment]
In a previous post in this thread I mentioned because of a conflict with `next-router` and semantic-ui-react I switched routers. But the errors shown are pointing to next's router. Is there a way to suppress the error? I wonder if I removed the `router` library?

**Reply by malixsys** [reply]
See my fix above ^^^

**Reply by antonioOrtiz** [reply]
Thank you for chiming in mate! Can I ask what version of `next` you are using?

**Reply by malixsys** [reply]
NP

```json
"next": "^9.4.4",
```

**Comment by antonioOrtiz** [comment]
So I went back and took a look at [next-react-router](https://github.com/toomuchdesign/next-react-router).  Essentially starting over.

And it appears originally I had this in my `pages/index file`

`import { Switch, Route, Redirect, withRouter } from 'react-router-dom';`

Instead what is [recommended.](https://github.com/toomuchdesign/next-react-router/blob/master/pages/index.js) 

```
import { Route, Link, Switch } from "react-router-dom";
import { Redirect } from 'react-router'
```

So with that in place, looks like that took me to this error:

`Unhandled Rejection (TypeError): Cannot read property 'originalUrl' of undefined
`
And it appears that was coming from this file [with-react-router](https://github.com/toomuchdesign/next-react-router/tree/master/next) in this part:

```
export default App => {
  return class AppWithReactRouter extends React.Component {
    static async getInitialProps(appContext) {
      const {
        ctx: {
          req: {
            originalUrl,
            locals = {},
          },
        },
      } = appContext;
      return {
        originalUrl,
        context: locals.context || {},
      };
    }
```

And after some console logging it looked like that was happening because of context or lack of it.

```
    static async getStaticProps(appContext) {
      const { ctx } = appContext;
      const { req } = ctx;
      const { locals } = req;

      console.log('req ', Object.keys(req));

      return {
        originalUrl: req.originalUrl,
        context: locals.context || {}
      };
    }
```
So the above fixes the previous error but yields this one! 

`Uncaught (in promise) TypeError: Cannot read property 'split' of undefined`

But after some digging looks like so someone ran into that using [next-router](https://github.com/vercel/next.js/issues/9698)

![Screen Shot 2020-06-29 at 2 53 55 PM](https://user-images.githubusercontent.com/4282013/86044590-47837380-ba18-11ea-9283-7ee01d2ed766.jpg)

At this point I am torn as to continue with this or ripping out `next-react-router` and go back to `next-router`, I actually did a little inspecting and found in my case I really just need something for `redirects` which I found [here](https://sergiodxa.com/articles/redirects-in-next-the-good-way/), And a component which I can use for my private routes on the client...In retrospect I wish I went with something like [next-authetication.](https://github.com/j0lv3r4/next-authentication#redirecturl-string-optional)

**Comment by EitanElbaz** [comment]
Our project is a SPA with a single entry point using a dynamic catch all file called `[...default].tsx` and ran into the same issue. https://nextjs.org/docs/routing/dynamic-routes#catch-all-routes

To get around it we used `beforePopState` from next/router and return false which is documented here https://nextjs.org/docs/api-reference/next/router#routerbeforepopstate

Because react-router-dom does all of our client side routing anyway we don't need Next to interfere.

**Comment by Infonautica** [comment]
Same issue. For private part of website i set "Catch all" route with React-Router v5 inside. Everything works great, except going back to previous page, where i get
```
Unhandled Runtime Error
TypeError: Cannot read property 'indexOf' of undefined

Call Stack
addBasePath
../../../../next-server/lib/router/router.ts (22:9)
```

**Comment by Timer** [comment, marked as answer]
Will everyone please try `next@^9.4.5-canary.36` and see if that fixes it for them? I'm going to mark this as accepted so it floats to the top, but let us know if this still reproduces.

**Reply by Infonautica** [reply]
canary.35 is latest available for me right now

**Reply by Timer** [reply]
That version should be new enough!

**Reply by Infonautica** [reply]
Here my error at canary.36
![image](https://user-images.githubusercontent.com/5521648/87688818-1e8d0d80-c790-11ea-8ddc-f25bfbb5e4dd.png)

**Reply by Timer** [reply]
Can you please open a new issue (not discussion) with reproduction? I'll leave this marked as answered since we fixed the original error in OP.

**Reply by weerayut601** [reply]
![image](https://user-images.githubusercontent.com/59607794/112322513-8d3d4600-8ce3-11eb-8b15-4768d0539523.png)
my error. How to fix it?
